README.md
# tiffy

## About
Generate Feeds from TIE Content.

## Requirements
### Base
- Python 3.7
- TIE API Key http://tie.dcso.de  

### Packages
- PyTest https://pytest.org
- pytest-testdox https://github.com/renanivo/pytest-testdox
- Requests http://python-requests.org
- PyYAML http://pyyaml.org
- Click http://click.pocoo.org/
- PyMISP https://github.com/MISP/PyMISP
- python-dateutil https://dateutil.readthedocs.io

## Install
```bash
$ git clone https://github.com/DCSO/tiffy.git
$ pip3 install -r requirements.txt
```

## Configuration
The command line client expects a configuration file in the `tiffy/settings` directory where you have to
define the required API key and URL. To create the config file, just copy the `config.sample.yml` file to `config.yml`
and edit it. You also need to copy `tags.sample.yml` and set up any desired tags to be set. 

```bash
$ cp settings/config.sample.yml settings/config.yml

$ vim settings/config.yml
```

## HowTo
To start the generator just run:
```bash
$ ./tiffy.py
```
The generator will now process all IOCs as attributes beginning from the current system date.

If no first seen date is set, the generator will always use the current system date as default.

To process attributes from or until a specific date you can use the `--first-seen YYYY-MM-DD` or 
`--last-seen YYYY-MM-DD` option. You can also combine both parameters.

```bash
$ ./tiffy.py --first-seen 2019-03-13

$ ./tiffy.py --last-seen 2019-07-13

$ ./tiffy.py --first-seen 2019-03-13 --last-seen 2019-07-13
```

### Using the Source, Actor, Category or Family parameter
Sometimes it's necessary to get all IOC's from a specific attacker group or tool family. In most cases these are known 
under more than one name. Because of this, tiffy offers the capability to search for these values.

You can pass one or multiple values either as a single string or a comma delimited list of strings

Query family based IOC's
```bash
$ ./tiffy.py --family example

$ ./tiffy.py --family example1,example2
```

Query actor based IOC's
```bash
$ ./tiffy.py --actor example

$ ./tiffy.py --actor example1,example2
```

Query category based IOC's
```bash
$ ./tiffy.py --category example

$ ./tiffy.py --category example1,example2
```

Query source based IOC's
```bash
$ ./tiffy.py --source example

$ ./tiffy.py --source example1,example2
```

## Using Severity and Confidence Parameters

tiffy is also able to filter ioc's based on min or max severity/confidence. If you pass only a min-value, tiffy
will search all values at or above the value. If you pass only a max-value, ioc's at or below the value will be returned.
When passing min and max-value, all ioc's between these values are used.

Query severity based IOC's
```bash
$ ./tiffy.py --min-severity 2                          #gets all ioc's from and including severity 2    

$ ./tiffy.py --max-severity 4                          #gets all ioc's up to and including severity 4

$ ./tiffy.py --min-severity 2 --max-severity 4         #gets all ioc's from severity 2 to severity 4
```

Query confidence based IOC's
```bash
$ ./tiffy.py --min-confidence 2                        #gets all ioc's from and including confidence 2    

$ ./tiffy.py --max-confidence 4                        #gets all ioc's up to and including confidence 4

$ ./tiffy.py --min-confidence 2 --max-confidence 4     #gets all ioc's from confidence 2 to confidence 4
```

## Query specific data types

You can specify the data type to be queried by TIE. This is useful, for example, if you use the txt output parameter to
create specific files that contain only domains, URL, and so on.

```bash
$ ./tiffy.py --data-type urlverbatim

$ ./tiffy.py --data-type urlverbatim,domainname

$ ./tiffy.py --data-type urlverbatim  --output-format txt
```

## Setting Default Tags for the MISP Event

You can pass tags for the newly created event. Tags are passed as MISP-compatible JSON Strings and will be added 
to the base event. Double quotes need to be escaped. If no tags are passed `TLP:AMBER` will be used as default.

```bash
$ ./tiffy.py --event-tags {\"name\":\"tlp:amber\"}
```

## Setting Tags for Attributes

You can assign Tags to attributes using the `settings/tags.yml` file. Currently only attributes derived from TIE
`c2-server` and `malware` observations can be tagged. Add your desired Tag to the list of tags as seen in the sample
file. The format for a new entry is:

```
- name: tag:name:here
```

## Setting the Output Format

You can choose the output format of the feed. Currently MISP-JSON and text files are supported but more formats will
follow.

```bash
$ ./tiffy.py --output-format MISP

$ ./tiffy.py --output-format txt
```

## Disable the Default Filter

tiffy will use the default TIE filter. You can disable this behaviour by passing the `--no-filter` parameter.

```bash
$ ./tiffy.py --no-filter
```

## Additional Parameters

tiffy offers some additional parameters:
- `--loglvl` sets the log level. Values are 0 - NOTSET / 10 - DEBUG / 20 - INFO / 30 - WARNING / 40 - ERROR / 50 - CRITICAL
- `--disable_console_log` disables log output to the console
- `--disable_file_log` disables logging to file
- `--log_file_path` sets the path where the file log should be saved

```bash
$ ./tiffy.py --loglvl 10

$ ./tiffy.py --disable_console_log

$ ./tiffy.py --disable_file_log

$ ./tiffy.py --log_file_path "path/to/log/destination"
```

## Using a Proxy
tiffy offers various ways for the use of a proxy. First, if the system variable `HTTP_PROXY` or `HTTPS_PROXY` is 
set, tiffy will automatically use the given information's.

If no system variable is used, tiffy will check if the parameter `--proxy_http` or `--proxy_https` is set. If so, tiffy will use the parameter for pulling informations. 

You can use only `--proxy_http` or `--proxy_https` or both
```bash
$ ./tiffy.py --proxy_http "http://10.8.0.1:8000"
$ ./tiffy.py --proxy_http "http://10.8.0.1:8000 --proxy_https "http://10.8.0.1:8443"
```
With HTTP Basic Auth
```bash
$ ./tiffy.py --proxy_http "http://user:pass@10.8.0.1:8000"
```

### Disable Certificate Verification

If your Proxy is using SSL-Interception, it might be necessary to disable
the certificate verification for requests. Use the `--disable_cert_verify` flag
in this case.

```bash
$ ./tiffy.py --proxy_http "http://10.8.0.1:8000" --disable_cert_verify
```

### Environment Variables

All config file settings can also be set using the environment variables documented in the tables below.
You can run tiffy without a config file only using environment variables, but in this case you MUST provide
at least the required variables.

#### Required Variables

| Variable                          | Default | Example                                | Description                    |
| --------------------------------- | ------- | -------------------------------------- | ------------------------------ |
| TIFFY_CONF_TIE_APIURL             |         |  https://tie.dcso.de/v1/api            | URL to TIE.                    |
| TIFFY_CONF_TIE_APIKEY             |         |  12345683127481209123789               | API token for TIE access       |
| TIFFY_CONF_MISP_ORGANISATION_NAME |         |  ACME                                  | Name of your MISP organization |
| TIFFY_CONF_MISP_ORGANISATION_UUID |         |  5804adw2-12fe-1234-34av-07lk82aw012a  | UUID of your MISP organization |

#### Optional Variables

| Variable                                 | Default              | Example                  | Description                                                       |
| ---------------------------------------- | -------------------- | ------------------------ | ----------------------------------------------------------------- |
| TIFFY_CONF_MISP_EVENTS_BASE_THREAT_LEVEL | 3                    |                          | IoC will get this threat level if it is added                     |
| TIFFY_CONF_MISP_EVENTS_BASE_CONFIDENCE   | 80                   |                          | IoC will get this confidence if it is added                       |
| TIFFY_CONF_MISP_EVENTS_BASE_SEVERITY     | 2                    |                          | IoC will get this severity if it is added                         |
| TIFFY_CONF_MISP_EVENTS_PUBLISHED         | false                |                          | IoC will get published in MISP                                    |
| TIFFY_CONF_MISP_ATTRIBUTES_TO_IDS        | false                |                          | Set IDS flag for this IoC                                         |
| TIFFY_CONF_MISP_ATTRIBUTES_TO_IDS_THRESHOLD| 90                 |  90                      | Confidence Threshold at or above which attributes should get IDS flag                                        |
| TIFFY_PARAM_TIE_SEEN_FIRST               |                      | YYYY-MM-DD               | Download only IoC which are first seen at ... and newer           |
| TIFFY_PARAM_TIE_SEEN_LAST                |                      | YYYY-MM-DD               | Download only IoC which are last seen at ... and older            |
| TIFFY_PARAM_TIE_ACTOR                    |                      | example1,example2        | Download only IoC with this actor                                 |
| TIFFY_PARAM_TIE_CATEGORY                 |                      | example1,example2        | Download only IoC with this category                              |
| TIFFY_PARAM_TIE_FAMILY                   |                      |   example1,example2      | Download only IoC with this family                                |
| TIFFY_PARAM_TIE_SOURCE                   |                      | example1,example2        | Download only IoC from this source                                |
| TIFFY_PARAM_TIE_SEVERITY_MIN             |                      | 2                        | Download only IoC with this minimum severity                      |
| TIFFY_PARAM_TIE_SEVERITY_MAX             |                      | 4                        | Download only IoC with this maximum severity                      |
| TIFFY_PARAM_TIE_CONFIDENCE_MIN           |                      | 2                        | Download only IoC with this minimum confidence                    |
| TIFFY_PARAM_TIE_CONFIDENCE_MAX           |                      | 4                        | Download only IoC with this maximum confidence                    |
| TIFFY_PARAM_TIE_MISP_EVENT_TAGS          | {\\"name\\":\\"tlp:amber\\"}| {\\"name\\":\\"tlp:amber\\"} | Tag Event with the defined tags                                   |
| TIFFY_PARAM_OUTPUT_FORMAT                | MISP                 |                          | You can choose the output format of the feed.                     |
| TIFFY_PARAM_TIE_DISABLE_DEFAULT_FILTER   | false                | true / false             | To disable the default TIE filter.                                |
| TIFFY_PARAM_LOG_LEVEL                    | INFO                 |                          | Define one of these log levels: DEBUG, INFO, WARNING, ERROR, CRITICAL |
| TIFFY_PARAM_LOG_DISABLE_CONSOLE          | false                | true / false             | Disables log output to stdout                                     |
| TIFFY_PARAM_LOG_DISABLE_FILE             |   false              | true / false             | Disables log output to file                                       |
| TIFFY_PARAM_LOG_FILE                     | "~/tiffy.log"        |                          | Define the log path                                               |

#### Proxy Variables

| Variable    | Default | Example                               | Description                              |
| ----------- | ------- | ------------------------------------- | ---------------------------------------- |
| HTTP_PROXY  |         |  http://10.8.0.1:8000                 | Set an Proxy server for HTTP connections |
| HTTPS_PROXY |         |  https://<user>:<pass>@10.8.0.1:8000  | Set Proxy server for HTTPS connections   |

## Setting up Feed
if tiffy ran successfully at least once, the directory `tiffy/feed` will be present. In this directory are all files needed for a MISP Feed. You need to upload these files onto a file server like nginx or apache.

We used nginx as an example. Upload the files into a directory on the nginx server and add a server configuration in the nginx.conf to make the directory accessible.

```
server {
		listen 8001;  #port the nginx should listen on and provide the feed
	    root /path/to/feed/directory;
	    	autoindex on;        # tells nginx to automatically index the files so that they can be accessed
	    location / {
	    	autoindex on;        # tells nginx to automatically index the files so that they can be accessed
	    }
	}
```

After setting up the file server, the feed can be added to MISP. From the main menu, go to `Sync Actions`-> `List Feeds`. In the menu on the left, select `Add Feed`. 

Set the feed to enabled and activate lookup and caching. Name the feed and the provider and set input source to Network. In the url field, add the url to the feed directory on your file server.

![alt text](https://raw.githubusercontent.com/DCSO/tiffy/master/images/add_feed.png "Add Feed")

After setting up the feed and enabling it, the events will be imported into MISP. To manually start an import, click the download icon on the TIE feed in the feed list. To see all events and import single events contained in the feed, click on the magnifying glass icon.

![alt text](https://raw.githubusercontent.com/DCSO/tiffy/master/images/options.png "Options")

## License

This software is released under a BSD 3-Clause license.
Please have a look at the LICENSE file included in the repository.

Copyright (c) 2019, DCSO Deutsche Cyber-Sicherheitsorganisation GmbH


CHANGELOG.md
# tiffy Changelog

### 1.0.4
    #### Changes
    - added support for TIE data types
    - added support to create txt files
    - added gitignore entry to ignore existing feed files

### 1.0.2
   #### Changes
   - added threshold setting for IDS flag in config
   - tagging for attributes derived from TIE observations added
   - removed unused config values base_severity and base_confidence
   
   #### Bugfixes
   - Fixed Bug with logging when no log_path was provided

### 1.0.1
   #### Changes
   - simplified Config File format, added possibility to use environment variables for configuration (see readme)
   - added parameter to change log saving location
   - loglvl now takes string values, e.g. DEBUG or INFO
   - removed unused config values and parameters (e.g. Attributes_Tagging)
   - added flag to diasble certificate validation (needed in some proxy setups)
   
   #### Bugfixes
   - Fixed Attributes to_ids config value not being used / not working
   - Fixed Event Published config value not being used / not working
   - Fixed http proxy not being used as https proxy if no specific https proxy was given
   - Fixed https proxy parameter unnecessarily enforcing https scheme
   
   

tiffy.py
#!/usr/bin/env python3

"""
# #####################################################################################
#                                                                                     #
#                                                                                     #
#                    .___________. __   _______  ___________    ____                  #
#                    |           ||  | |   ____||   ____\   \  /   /                  #
#                    `---|  |----`|  | |  |__   |  |__   \   \/   /                   #
#                        |  |     |  | |   __|  |   __|   \_    _/                    #
#                        |  |     |  | |  |     |  |        |  |                      #
#                        |__|     |__| |__|     |__|        |__|                      #
#                                                                                     #
#                                                                                     #
#                                                                                     #
# #####################################################################################

DCSO tiffy

Copyright (c) 2019, DCSO GmbH
Project   - https://github.com/DCSO/

"""
import json
import logging
import os
import re
import signal as signal_module
import sys
from datetime import datetime
from json import JSONDecodeError
from urllib.parse import urlparse

import click
import pytest
from click.testing import CliRunner

from TIELoader import TIELoader
from model import Config, Tags


# CTRL+C Handler --------------------------------------------------------------
def signal_handler(signal_name, frame):
    try:
        logging.info("------------------------------------------\n")
        logging.info('tiffy has been interrupted. Shutting down flux capacitor.')
    except Exception as e:
        print('tiffy has been interrupted. Shutting down flux capacitor.')
    sys.exit(0)


# INIT tiffy --------------------------------------------------------------
@click.command()
@click.option('--category', help='list of categories separated by comma to filter for',
              envvar='TIFFY_PARAM_TIE_CATEGORY')
@click.option('--actor', help='list of actors separated by comma to filter for', envvar='TIFFY_PARAM_TIE_ACTOR')
@click.option('--family', help='list of actors separated by comma to filter for', envvar='TIFFY_PARAM_TIE_FAMILY')
@click.option('--source', help='list of source pseudonyms separated by comma to filter for',
              envvar='TIFFY_PARAM_TIE_SOURCE')
@click.option('--first-seen', envvar='TIFFY_PARAM_TIE_SEEN_FIRST')
@click.option('--last-seen', envvar='TIFFY_PARAM_TIE_SEEN_LAST')
@click.option('--event-tags', help='event base tags as MISP conform JSON object String',
              default='{"name": "tlp:amber"}', envvar='TIFFY_PARAM_TIE_MISP_EVENT_TAGS')
@click.option('--output-format', type=click.Choice(['MISP', "txt"]), default='MISP',
              help='sets the output format for the feed.', envvar='TIFFY_PARAM_OUTPUT_FORMAT')
@click.option('--no-filter', is_flag=True,
              help='If set, no filter will be used for the request to TIE. Otherwise, the default filter will be used.',
              envvar='TIFFY_PARAM_TIE_DISABLE_DEFAULT_FILTER')
@click.option('--loglvl', default='INFO', help='Sets the log level. Default is INFO.\n Params are: DEBUG / '
                                               'INFO / WARNING / ERROR / CRITICAL', envvar='TIFFY_PARAM_LOG_LEVEL')
@click.option('--disable_console_log', is_flag=True, help='If used, the convert will not generate output in the '
                                                          'console', envvar='TIFFY_PARAM_LOG_DISABLE_CONSOLE')
@click.option('--disable_file_log', is_flag=True, help='If used, the converter will not generate a file output',
              envvar='TIFFY_PARAM_LOG_DISABLE_FILE')
@click.option('--log_file_path', help='path to the location where the log file should be stored',
              envvar='TIFFY_PARAM_LOG_FILE')
@click.option('--min-severity', type=int, help='Events with a severity value lower than the given value will not be '
                                               'fetched. If used, the generator will ignore the severity value defined '
                                               'in the config file. The severity value can be equal or between 0 and 5.',
              envvar='TIFFY_PARAM_TIE_SEVERITY_MIN')
@click.option('--min-confidence', type=int,
              help='Events with a confidence value lower than the given value will not be '
                   'fetched. If used, the generator will ignore the severity value defined '
                   'in the config file. The confidence value can be equal or between 0 and '
                   '100', envvar='TIFFY_PARAM_TIE_CONFIDENCE_MIN')
@click.option('--max-severity', type=int, help='Events with a severity value higher than the given value will not be '
                                               'fetched. If used, the generator will ignore the severity value defined '
                                               'in the config file. The severity value can be equal or between 0 and 5.',
              envvar='TIFFY_PARAM_TIE_SEVERITY_MAX')
@click.option('--max-confidence', type=int,
              help='Events with a confidence value higher than the given value will not be '
                   'fetched. If used, the generator will ignore the severity value defined '
                   'in the config file. The confidence value can be equal or between 0 and '
                   '100', envvar='TIFFY_PARAM_TIE_CONFIDENCE_MAX')
@click.option('--data-type', help='list of data types supported by TIE separated by comma')
@click.option('--proxy_http', type=str, help='Sets the address for a http based proxy e.g. http://10.8.0.1:8000')
@click.option('--proxy_https', type=str, help='Sets the address for a https based proxy e.g. https://10.8.0.1:8000')
@click.option('--disable_cert_verify', is_flag=True,
              help='If set, ssl-certs will not be validated.')
def init(category, actor, family, source, first_seen, last_seen, event_tags, output_format, no_filter, loglvl,
         disable_console_log, disable_file_log, log_file_path, min_severity, min_confidence, max_severity,
         max_confidence, data_type, proxy_http, proxy_https, disable_cert_verify):
    """
    Starting the converter
    """
    error = False
    given_first_seen_date = ''
    given_last_seen_date = ''
    proxy_tie_addr = dict()

    # Signal handler for CTRL+C
    signal_module.signal(signal_module.SIGINT, signal_handler)
    try:
        event_tags = json.loads(event_tags)
    except JSONDecodeError:
        raise_error_critical('event tags are not valid JSON')

    # check loglvl
    if isinstance(loglvl, str):
        if loglvl not in ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']:
            click.echo(
                'loglvl must be one of DEBUG, INFO, WARNING, ERROR or CRITICAL. Type \'python tiffy.py --help\' for'
                ' more information\'s.\nSet default value: INFO')
            loglvl = 'INFO'
    else:
        click.echo('loglvl must be an unsigned integer value equal or between 0 and 50. Type \'python tiffy.py '
                   '--help\' for more information\'s.\nSet default value: 20')
        loglvl = 'INFO'
    if log_file_path is None:
        log_file_path = sys.path[0]
    TIELoader.init_logger(log_file_path, "tiffy.py.log", loglvl, disable_console_log, disable_file_log)
    try:

        # Check date arguments
        if first_seen is not None:
            try:
                given_first_seen_date = datetime.strptime(first_seen, "%Y-%m-%d")
            except ValueError:
                raise_error_critical(
                    'First Seen Date could not be converted. Please use the following format YYYY-MM-DD')
        else:
            given_first_seen_date = datetime.now()

        # Check date arguments
        if last_seen is not None:
            try:
                given_last_seen_date = datetime.strptime(last_seen, "%Y-%m-%d")
            except ValueError:
                raise_error_critical(
                    'Last Seen Date could not be converted. Please use the following format YYYY-MM-DD')

        # Check confidence

        if max_confidence is not None:
            if isinstance(max_confidence, int):
                if max_confidence < 0 or max_confidence > 100:
                    raise_error_critical(
                        'The max confidence value must be an unsigned integer value equal or between 0 and 100')
            else:
                raise_error_critical('The max confidence value must be an unsigned integer value')
        if min_confidence is not None:
            if isinstance(min_confidence, int):
                if min_confidence < 0 or min_confidence > 100:
                    raise_error_critical(
                        'The confidence value must be an unsigned integer value equal or between 0 and 100')
                elif max_confidence and (min_confidence > max_confidence):
                    raise_error_critical(
                        'The min confidence must be lower than max confidence')
            else:
                raise_error_critical('The confidence value must be an unsigned integer value')

        # Check severity
        if max_severity is not None:
            if isinstance(max_severity, int):
                if max_severity < 0 or max_severity > 5:
                    raise_error_critical(
                        'The max severity value must be an unsigned integer value equal or between 0 and 5')
            else:
                raise_error_critical('The max severity value must be an unsigned integer value')
        if min_severity is not None:
            if isinstance(min_severity, int):
                if min_severity < 0 or min_severity > 5:
                    raise_error_critical(
                        'The severity value must be an unsigned integer value equal or between 0 and 5')
                elif max_severity and (min_severity > max_severity):
                    raise_error_critical(
                        'The min severity must be lower than max severity')
            else:
                raise_error_critical('The severity value must be an unsigned integer value')

        # Check Proxy variables
        ## First check if a Proxy for TIE has been set
        if proxy_http is not None or os.environ.get('HTTP_PROXY') or proxy_https is not None or os.environ.get(
                'HTTPS_PROXY'):
            proxy_tie_addr = checkProxyUrls(proxy_http, proxy_https, True)

        # check family, source, category, actor parameters
        pattern = "^[a-zA-Z0-9 /-]+(?:,[a-zA-Z0-9 /-]+)*$"
        if actor:
            if not re.search(pattern, actor):
                raise_error_critical(
                    'actor must be a single value or a comma delimited list of values')
        if family:
            if not re.search(pattern, family):
                raise_error_critical(
                    'family must be a single value or a comma delimited list of values')
        if category:
            if not re.search(pattern, category):
                raise_error_critical(
                    'category must be a single value or a comma delimited list of values')
        if source:
            if not re.search(pattern, source):
                raise_error_critical(
                    'source must be a single value or a comma delimited list of values')

        #### everything is fine -> start up
        # Loading tags file
        try:
            attr_tags = Tags.parse("settings/tags.yml")
        except FileNotFoundError:
            attr_tags = Tags()
        try:
            # Load config
            conf = Config.parse("settings/config.yml")

            logging.info("Powering up flux capacitor. Starting up tiffy.")
            logging.info("#### Start new TIE-Query ####")
            logging.debug("Params:")
            logging.debug("output_format: " + str(output_format))
            logging.debug("event_tags: " + str(event_tags))
            logging.debug("category: " + str(category))
            logging.debug("actor: " + str(actor))
            logging.debug("family: " + str(family))
            logging.debug("source: " + str(source))
            logging.debug("given_first_seen_date: " + str(given_first_seen_date))
            logging.debug("given_last_seen_date: " + str(given_last_seen_date))
            logging.debug("min_confidence: " + str(min_confidence))
            logging.debug("min_severity: " + str(min_severity))
            logging.debug("max_confidence: " + str(max_confidence))
            logging.debug("max_severity: " + str(max_severity))
            logging.debug("conf.event_base_thread_level: " + str(conf.event_base_thread_level))
            logging.debug("conf.base_severity: " + str(conf.base_severity))
            logging.debug("conf.event_published: " + str(conf.event_published))
            logging.debug("conf.tie_api_key: " + str(conf.tie_api_key))
            logging.debug("conf.tie_api_url: " + str(conf.tie_api_url))
            logging.debug("conf.attr_to_ids: " + str(conf.attr_to_ids))
            logging.debug("conf.base_confidence: " + str(conf.base_confidence))
            logging.debug("conf.org_uuid: " + str(conf.org_uuid))
            logging.debug("conf.org_name: " + str(conf.org_name))
            logging.debug("log_file_path: " + str(log_file_path))

            TIELoader.start(output_format, conf, event_tags, attr_tags, category, actor, family, source, given_first_seen_date,
                            given_last_seen_date, min_confidence, min_severity, max_confidence, max_severity,
                            proxy_tie_addr, data_type, no_filter, disable_cert_verify)

        except FileNotFoundError:
            logging.error("Error: \nconfig.yml not found")

    except (RuntimeError, TypeError) as ex:
        click.echo(ex)


def checkProxyUrls(proxy_http, proxy_https, system_proxy=True):
    url_http = None
    url_https = None
    proxy_addrs = dict()

    if proxy_http is not None:
        url_http = urlparse(proxy_http)
    elif system_proxy:
        if os.environ.get('HTTP_PROXY'):
            url_http = urlparse(os.environ['HTTP_PROXY'])
    if proxy_https is not None:
        url_https = urlparse(proxy_https)
    elif proxy_http is not None:
        url_https = urlparse(proxy_http)
    elif system_proxy:
        if os.environ.get('HTTPS_PROXY'):
            url_https = urlparse(proxy_https)

    # check if HTTP attributes are valid
    if url_http is not None:
        if url_http.scheme is None or url_http.port is None or url_http.hostname is None:
            raise_error_critical(
                'HTTP Proxy address ist not valid. Type \'python tiffy.py --help\' for more information\'s.')
        if url_http.scheme != 'http':
            raise_error_critical('HTTP Proxy address must have a valid scheme')
        if url_http.port <= 0 or url_http.port > 65535:
            raise_error_critical('HTTP Proxy address must have a valid port')
        if len(url_http.hostname) <= 2:
            raise_error_critical('HTTP Proxy address is to short or not valid ')
        # Address should be valid
        proxy_addrs['http'] = str(url_http.scheme + "://" + url_http.netloc)
        # If not set, Request an PyMISP will not querry HTTPS Urls
        proxy_addrs['https'] = str(url_http.scheme + "://" + url_http.netloc)

    # check if HTTPS attributes are valid
    if url_https is not None:
        if url_https.scheme is None or url_https.port is None or url_https.hostname is None:
            raise_error_critical(
                'HTTPS Proxy address is not valid. Type \'python tiffy.py --help\' for more information\'s.')
        if url_https.scheme != 'https' and url_https.scheme != 'http':
            raise_error_critical('HTTPS Proxy address must have a valid scheme')
        if url_https.port <= 0:
            raise_error_critical('HTTP Proxy address must have a valid port')
        if len(url_https.hostname) <= 2:
            raise_error_critical('HTTP Proxy address is to short or not valid ')
        # Address should be valid
        proxy_addrs['https'] = str(url_https.scheme + "://" + url_https.netloc)

    return proxy_addrs


def raise_error_critical(error_str):
    ERROR_BASE_STR = "Error starting tiffy.py: "
    logging.error(ERROR_BASE_STR + error_str)
    raise RuntimeError(ERROR_BASE_STR + error_str)


def raise_error_warning(error_str):
    ERROR_BASE_STR = "Error starting tiffy.py: "
    logging.warning(ERROR_BASE_STR + error_str)


# MAIN ################################################################
if __name__ == '__main__':
    init()
